### DISCUSSION TITLE
implementation of over segmentation algorithms ?

### DISCUSSION THREAD

**Discussion by QuanticDisaster** [pyg-team/pytorch_geometric#2715] [Ideas]
Hello,

I was looking for over-segmentation algorithms on 3D points clouds, and while I encountered a few papers about it, finding an implementation in popular libraries seems quite hard. Specifically, I was thinking about segmentation algorithms respecting objects and geometry so that a region doesn't cross on two soon-to-be classified objects to use in semantic segmentation field. Examples of the papers I encountered :
https://arxiv.org/pdf/1702.04114.pdf : Graph Based Over-Segmentation Methods for 3D PointClouds
https://arxiv.org/pdf/1711.09869.pdf : Large-scale Point Cloud Semantic Segmentation with Superpoint Graphs

I was wondering if pytorch was planning to add such features or if I missed a library/function doing that. The goal would be to smooth or average predictions on per-points segmentation

**Comment by rusty1s** [comment]
Thanks for the pointers! You are right that current PyG lacks some more examples regarding point cloud processing and segmentation in particular. We might wanna add those models (or related ones) to the library in the near future. If you are interested in contributing, please feel free to reach out :)

**Reply by QuanticDisaster** [reply]
Happy  to hear that ! I fear that some of the mathematics and understanding are beyond me to try them myself.

As a roundabout for those interesed, I tried to draw inspiration from the local_smoothing and graph_cut methods of CGAL library : 
https://doc.cgal.org/latest/Classification/index.html#Classification_classification_functions
https://github.com/CGAL/cgal/blob/master/Classification/include/CGAL/Classification/classify.h

However, as the energy defined here correspond to the score of each class for points which is between 0 and 1 in their case, we need a normalization method for neural networks who outputs probability on the range of R
As a consequence I just used an averaging on the class of the neighboors to smooth the resulting point_cloud for the moment
